includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-update.php

```diff
@@ -0,0 +1,271 @@
+
+			// Update the subscription with the updated SEPA gateway ID.
+			$this->set_subscription_updated_payment_method( $subscription );
+
+			$this->log( sprintf( 'Successful migration of subscription #%1$d.', $subscription_id ) );
+		} catch ( \Exception $e ) {
+			$this->log( $e->getMessage() );
+		}
+	}
+
+	/**
+	 * Gets the subscription to update.
+	 *
+	 * Only allows migration if:
+	 * - The Legacy experience is disabled
+	 * - The WooCommerce Subscription extension is active
+	 * - The subscription ID is a valid subscription
+	 *
+	 * @param int $subscription_id The ID of the subscription to migrate.
+	 * @return WC_Subscription The Subscription object for which its token must be updated.
+	 * @throws \Exception Skip the migration if the request is invalid.
+	 */
+	private function get_subscription_to_migrate( $subscription_id ) {
+		if ( ! WC_Stripe_Feature_Flags::is_upe_checkout_enabled() ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. The Legacy experience is enabled.', $subscription_id ) );
+		}
+
+		if ( ! class_exists( 'WC_Subscriptions' ) ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. The WooCommerce Subscriptions extension is not active.', $subscription_id ) );
+		}
+
+		$subscription = wcs_get_subscription( $subscription_id );
+
+		if ( ! $subscription ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. Subscription not found.', $subscription_id ) );
+		}
+
+		return $subscription;
+	}
+
+	/**
+	 * Returns an updated token to be used for the subscription, given the source ID.
+	 *
+	 * If no updated token is found, we create a new one based on the legacy one.
+	 *
+	 * @param string  $source_id The Source or Payment Method ID associated with the subscription.
+	 * @param integer $user_id   The WordPress User ID to whom the subscription belongs.
+	 */
+	private function maybe_create_updated_sepa_token_by_source_id( string $source_id, int $user_id ) {
+
+		// Retrieve the updated SEPA tokens for the user.
+		$replacement_token = $this->get_customer_token_by_source_id( $source_id, $user_id, $this->updated_sepa_gateway_id );
+
+		// If no updated SEPA token was found, create a new one based on the source ID.
+		if ( ! $replacement_token ) {
+			$replacement_token = $this->create_updated_sepa_token( $source_id, $user_id );
+		}
+	}
+
+	/**
+	 * Get the token for the user by its source ID and gateway ID. s
+	 *
+	 * @param string  $source_id  The ID of the source we're looking for.
+	 * @param integer $user_id    The ID of the user we're retrieving tokens for.
+	 * @param string  $gateway_id The ID of the gateway of the tokens we want to check.
+	 *
+	 * @return WC_Payment_Token|false
+	 */
+	private function get_customer_token_by_source_id( string $source_id, int $user_id, string $gateway_id ) {
+		$customer_tokens = WC_Payment_Tokens::get_customer_tokens( $user_id, $gateway_id );
```

Comment:
I wonder if we should introduce a helper function for this kind of thing. I think I had to do something similar (query for a token by `pm_`), and WC doesn't support that. 

I other reason I ask is because there's the potential that if the customer has a lot of tokens (more than 10) this function may fail to locate the token because by default WC limits that query to 10. If that happens we then create a new token. Which will just further contribute to that problem.

Reply:
> I wonder if we should introduce a helper function for this kind of thing. I think I had to do something similar (query for a token by pm_), and WC doesn't support that.

Sounds great, yeah. I was surprised we didn't have a function to find a token by its `token` value.

> I other reason I ask is because there's the potential that if the customer has a lot of tokens (more than 10) this function may fail to locate the token because by default WC limits that query to 10. If that happens we then create a new token. Which will just further contribute to that problem.

Right. I tried to reduce that risk by retrieving only the tokens for a specific gateway instead of retrieving all tokens. This way, we'd have to reach the query limit for tokens for this gateway for tokens to be left out. It's less likely that a user will have 10 SEPA tokens saved, but yes, still possible. 

As an alternative to reduce that possibility, we could also use the `woocommerce_get_customer_payment_tokens_limit` [filter](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-payment-tokens.php#L84C37-L84C82) to increase that limit. Although I agree a helper function would actually fix the potential problem.

Would this be a blocker for the PR?